flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerTest.java

```diff
@@ -940,6 +955,177 @@ public void testCancelWithSavepoint() throws Exception {
+				Option.apply("tm"),
+				true,
+				TestingTaskManager.class);
+
+			ActorGateway taskManager = new AkkaActorGateway(taskManagerRef, leaderId);
+
+			// Wait until connected
+			Object msg = new TestingTaskManagerMessages.NotifyWhenRegisteredAtJobManager(jobManager.actor());
+			Await.ready(taskManager.ask(msg, askTimeout), askTimeout);
+
+			JobCheckpointingSettings snapshottingSettings = new JobCheckpointingSettings(
+				Collections.singletonList(sourceVertex.getID()),
+				Collections.singletonList(sourceVertex.getID()),
+				Collections.singletonList(sourceVertex.getID()),
+				checkpointInterval,
+				3600000,
+				0,
+				Integer.MAX_VALUE,
+				ExternalizedCheckpointSettings.none(),
+				null,
+				true);
+
+			jobGraph.setSnapshotSettings(snapshottingSettings);
+
+			// Submit job graph
+			msg = new JobManagerMessages.SubmitJob(jobGraph, ListeningBehaviour.DETACHED);
+			Await.ready(jobManager.ask(msg, askTimeout), askTimeout);
+
+			// Wait for all tasks to be running
+			msg = new TestingJobManagerMessages.WaitForAllVerticesToBeRunning(jobGraph.getJobID());
+			Await.ready(jobManager.ask(msg, askTimeout), askTimeout);
+
+			// ----------------------------------------------------------------
+			// Super ugly... sorry! But this is one of the few bad options
+			// to test this here. Ideally, we would have a factory that we
+			// can set in tests as desired. But we don't. So here we go...
+			msg = new RequestExecutionGraph(jobGraph.getJobID());
+			Object result = Await.result(jobManager.ask(msg, askTimeout), askTimeout);
+
+			ExecutionGraph eg;
+			if (result instanceof ExecutionGraphFound) {
+				// Sorry...
+				eg = (ExecutionGraph) ((ExecutionGraphFound) result).executionGraph();
+			} else {
+				throw new RuntimeException("Could not access ExecutionGraph for job with "
+					+ "ID " + jobGraph.getJobID() + ". Response: " + result.toString());
+
+			}
+
+			Field field = eg.getClass().getDeclaredField("checkpointCoordinator");
+			field.setAccessible(true);
+			CheckpointCoordinator coord = (CheckpointCoordinator) field.get(eg);
+			CheckpointCoordinator spiedCoord = Mockito.spy(coord);
+			field.set(eg, spiedCoord);
+			// ----------------------------------------------------------------
+
+			// Cancel with savepoint
+			msg = new JobManagerMessages.CancelJobWithSavepoint(jobGraph.getJobID(),
+				savepointTarget.getAbsolutePath());
+			CancellationResponse cancelResp = (CancellationResponse) Await.result(
+				jobManager.ask(msg, askTimeout), askTimeout);
+
+			if (cancelResp instanceof CancellationFailure) {
+				if (enablePeriodicCheckpoints) {
+					// Verify checkpoint scheduler deactivated and reactivated.
+					// A call to start checkpoint scheduler calls stop scheduler
+					// again. Therefore, we verify two calls for stop. Since we
+					// spy (I know...) on the coordinator after the job has
+					// started, we don't count calls before spying.
+					verify(spiedCoord, times(1)).startCheckpointScheduler();
```

Comment:
The thing is that the stopping of the scheduler is part of the expected behaviour of cancel-with-job-savepoint, because we don't want any checkpoints between the savepoint and cancel job (https://issues.apache.org/jira/browse/FLINK-4717). I think for that we do need the spying :-( It was simply not fully tested before... Does this make sense or am I missing your point?